packages/ember/tests/helpers/link_to_test/link_to_with_query_params_test.js

```diff
@@ -1,506 +1,567 @@
-      this.route('create');
+      this.runTask(() => {
+        equal(router.currentRouteName, 'cars.index', 'The active route is still cars');
+        equal(router.get('url'), '/cars?page=2', 'The url has been updated');
+        equal(carsController.get('page'), 2, 'The query params have been updated');
+      });
+    });
+  }
+
+  ['@test the {{link-to}} applies activeClass when query params are not changed']() {
+    this.addTemplate('index', `
+      {{#link-to (query-params foo='cat') id='cat-link'}}Index{{/link-to}}
+      {{#link-to (query-params foo='dog') id='dog-link'}}Index{{/link-to}}
+      {{#link-to 'index' id='change-nothing'}}Index{{/link-to}}
+    `);
+    this.addTemplate('search', `
+      {{#link-to (query-params search='same') id='same-search'}}Index{{/link-to}}
+      {{#link-to (query-params search='change') id='change-search'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' archive=true) id='same-search-add-archive'}}Index{{/link-to}}
+      {{#link-to (query-params archive=true) id='only-add-archive'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' archive=true) id='both-same'}}Index{{/link-to}}
+      {{#link-to (query-params search='different' archive=true) id='change-one'}}Index{{/link-to}}
+      {{#link-to (query-params search='different' archive=false) id='remove-one'}}Index{{/link-to}}
+      {{outlet}}
+    `);
+    this.addTemplate('search.results', `
+      {{#link-to (query-params sort='title') id='same-sort-child-only'}}Index{{/link-to}}
+      {{#link-to (query-params search='same') id='same-search-parent-only'}}Index{{/link-to}}
+      {{#link-to (query-params search='change') id='change-search-parent-only'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' sort='title') id='same-search-same-sort-child-and-parent'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' sort='author') id='same-search-different-sort-child-and-parent'}}Index{{/link-to}}
+      {{#link-to (query-params search='change' sort='title') id='change-search-same-sort-child-and-parent'}}Index{{/link-to}}
+      {{#link-to (query-params foo='dog') id='dog-link'}}Index{{/link-to}}
+    `);
+
+    this.router.map(function() {
+      this.route('search', function() {
+        this.route('results');
+      })
     });
-  });
 
-  App.CarsController = Controller.extend({
-    queryParams: ['page'],
-    page: 1
-  });
+    this.add('controller:search', Controller.extend({
+      queryParams: ['search', 'archive'],
+      search: '',
+      archive: false
+    }));
+
+    this.add('controller:search.results', Controller.extend({
+      queryParams: ['sort', 'showDetails'],
+      sort: 'title',
+      showDetails: true
+    }));
+
+    this.visit('/').then(() => {
+      let router = this.appRouter;
+      //Basic tests
+      this.shouldNotBeActive('#cat-link');
+      this.shouldNotBeActive('#dog-link');
+      run(router, 'handleURL', '/?foo=cat');
```

Comment:
Again this can chain `visit` calls :-) All these `handleURL` bits can be replaced with `visit`.